sys/ecc: add golay(24,12) half rate code

PeterKietzmann opened the pull request:
### Contribution description
This PR adds a *Golay(24,12)* en/decoder to `ecc`. The code is based on  the [`liquid-dsp` library](https://github.com/jgaeddert/liquid-dsp) . I applied **a lot** of changes and optimizations which is why I decided **against** a `pkg`


### Testing procedure
From the RIOT root directory type:
`make -C tests/unittests tests-ecc all term`


### Issues/PRs references
Based on #9988

smlng commented on sys/ecc/golay2412.c:
```diff
@@ -0,0 +1,416 @@
+    unsigned int num_bits_in = _dec_msg_len*8;
+
+    /* compute total number of blocks: ceil(num_bits_in/_m) */
+    unsigned int num_blocks = num_bits_in / _m + (num_bits_in%_m ? 1 : 0);
```
add spaces around module `%` and also put parentheses around ternary condition, also see Codacy report.

smlng commented on sys/ecc/golay2412.c:
```diff
@@ -0,0 +1,416 @@
+    unsigned int num_bits_out = num_blocks * _k;
+
+    /* compute total number of bytes out: ceil(num_bits_out/8) */
+    unsigned int num_bytes_out = num_bits_out/8 + (num_bits_out%8 ? 1 : 0);
```
dito, see above

smlng commented on sys/ecc/golay2412.c:
```diff
@@ -0,0 +1,416 @@
+    0x0d1d, 0x0a3b, 0x0477, 0x0ffe};
+
+/* parity check matrix [12 x 24] */
+static const unsigned int golay2412_H[12] = {
```
I'm unsure if `unsigned int` is safe here, when using non-32Bit platforms. Maybe make this `uint32_t`? and above use `uint16_t`? Need to check, how eg. Arduino likes this.

smlng commented:
building on macOS for atmega-2560 fails with:

```
/RIOT/sys/ecc/golay2412.c:71:5: error: unsigned conversion from 'long int' to 'unsigned int' changes value from '8390893' to '2285' [-Werror=overflow]
     0x008008ed, 0x004001db, 0x002003b5, 0x00100769,
```

which is one of many of such errors

smlng commented:
I compiled `BOARD=arduino-mega2560 make -C tests/unittests/ tests-ecc`

smlng commented:
this doesn't show up in murdock, bc it does not build unittests for atmega based boards 😞

PeterKietzmann commented:
@smlng mind to have an other look at my latest commits? I would like to
- get #9988 merged and rebase this PR
- squash everything nicely together and run Murdock
- wait until Monday so I can test on 8-bit and 16-bit hardware

PeterKietzmann commented:
@smlng now that #9988 was merged, we can proceed here. Would you mind to have a look at the latest commits. Did I address all your comments?

PeterKietzmann commented:
@smlng ping! Can I squash?

smlng commented:
you can squash. Though Arduino now compiles it still fails on the `arduino-mega2560`:

```
2018-10-01 13:40:43,262 - INFO # ...
2018-10-01 13:40:43,359 - INFO # EccTest.test_golay2412_message (tests/unittests/tests-ecc/tests-ecc.c 100) exp 0 was -1
2018-10-01 13:40:43,359 - INFO # 
2018-10-01 13:40:43,359 - INFO # run 3 failures 1
```

PeterKietzmann commented:
@smlng it works with #10102

miri64 commented:
Please rebase :-).

PeterKietzmann commented:
Rebased, changed to the new 32-bit `bitarithm` function (compare #10102) and immediately squashed. *tests-ecc* succeeds on samr21-xpro (32-bit), z1 (16-bit) and arduino-mega2560 (8-bit) platforms.

miri64 commented on sys/ecc/golay2412.c:
```diff
@@ -351,7 +351,9 @@ void golay2412_encode(
 
-    assert( j== block_get_enc_msg_len(_dec_msg_len,12,24) );
+    if(j != block_get_enc_msg_len(_dec_msg_len,12,24)) {
+        assert(!"j != block_get_enc_msg_len(_dec_msg_len,12,24)");
```
Sorry, but this contradicts all benefits of `assert()`. If you get an `unused function` error because you only use this function within the `assert()`, please `#ifndef` that function with `#ifndef NDEBUG` or if you want to use it with an if, use a proper return code.

PeterKietzmann replied:
Ah, `NDEBUG` was the variable I was missing. Will do and directly squash, okay? Wanted to run Murdock...

miri64 replied:
:+1: but please also consider my comments below ;-)

miri64 commented on sys/ecc/golay2412.c:
Same here

miri64 commented on sys/ecc/golay2412.c:
```diff
@@ -0,0 +1,433 @@
+}
+#endif
+
+static inline uint32_t block_get_enc_msg_len(uint32_t _dec_msg_len,
```
This looks rather long for an inline function. Are you sure you intentended it to be inline?

miri64 commented on sys/ecc/golay2412.c:
```diff
@@ -0,0 +1,433 @@
+}
+
+/* multiply input vector with parity check matrix, H */
+static inline uint32_t golay2412_matrix_mul(uint32_t   _v,
```
Dito

miri64 commented on sys/ecc/golay2412.c:
Dito

miri64 commented on sys/ecc/golay2412.c:
And so on.

PeterKietzmann commented:
Oh dear, this is an endless story. I'll give it a try.

PeterKietzmann commented:
@miri64 here we go. Agree to squash and finally run Murdock again?

miri64 commented:
I'll have a last look at the PR

miri64 commented on tests/unittests/tests-ecc/tests-ecc.c:
```diff
@@ -75,11 +76,36 @@ static void test_hamming256_padding(void)
+    msg_enc[6]  ^= (1L << 16) | (1L << 22) | (1L << 19);
+    msg_enc[9]  ^= (1L << 11) | (1L <<  8) | (1L << 15);
+    msg_enc[12] ^= (1L <<  9) | (1L <<  2) | (1L << 10);
+    msg_enc[15] ^= (1L << 17) | (1L <<  1) | (1L <<  7);
```
Some interesting "corner-case" maybe would here to also have only 2 errors or 1 error or no errors not always 3 ;-).

PeterKietzmann commented:
@miri I changed the return values to `asserts`. The original implementation uses `exit(1)` so I guess this should be fine. Furthermore I extended the unit test and applied two minor changes. Tests succeed again on all three aforementioned platforms.

miri64 commented:
Right... the unittests don't fit `samr21-xpro` anymore... let me quickly check it on some boards in iot-lab.

miri64 commented:
Tested `tests-ecc` of `unittests` on `arduino-zero`, `iotlab-m3`, `samr21-xpro`, and `wsn430-v1_3b`. ACK, please squash.

miri64 commented on sys/include/ecc/golay2412.h:
```diff
@@ -0,0 +1,77 @@
+/**
+ * @brief encode block of data using Golay(24,12) encoder
+ *
+ * @pre _dec_msg_len and sizeof(_msg_dec) accirdingly have to be multiples of 3
```
Ah I could also use this new GitHub feature ;-)

```suggestion
 * @pre _dec_msg_len and sizeof(_msg_dec) accordingly have to be multiples of 3
```
